File: 01/sol.py
def read_file(filename="input.txt"):
    with open("input.txt", "r") as f:
        return f.read().strip()


def solve1(input_string):
    blocks = input_string.split("\n\n")
    lines = [block.split("\n") for block in blocks]
    nums = [[int(line) for line in block] for block in lines]
    sums = [sum(block) for block in nums]
    return max(sums)


def solve2(input_string):
    blocks = input_string.split("\n\n")
    lines = [block.split("\n") for block in blocks]
    nums = [[int(line) for line in block] for block in lines]
    sums = [sum(block) for block in nums]
    sort = sorted(sums, reverse=True)
    return sum(sort[:3])


if __name__ == "__main__":
    input_string = read_file("input.txt")
    assert input_string is not None
    res1 = solve1(input_string)
    print(res1)
    res2 = solve2(input_string)
    print(res2)

File: 01/sol.jl
function main()
    filename = isempty(ARGS) ? "input.txt" : ARGS[1]
    input = rstrip(read(filename, String))
    x = parseinput(input)
    r1 = solve1(x)
    println(r1)
    r2 = solve2(x)
    println(r2)
end

function parseinput(input)
    blocks = split(input, "\n\n")
    lines = map(block -> split(block, "\n"), blocks)
    nums = map(block -> map(line -> parse(Int, line), block), lines)
    return nums
end

function solve1(x)
    sums = map(sum, x)
    return maximum(sums)
end

function solve2(x)
    sums = map(sum, x)
    return sum(sort(sums, rev=true)[1:3])
end

main()
